libs/java/server_common/src/main/java/com/yahoo/athenz/common/server/util/config/ConfigManager.java

```diff
@@ -0,0 +1,463 @@
+ * Also, {@link #getConfigValue} uses the system-properties as fallback in case a property is
+ *  not set in the config-sources (using {@link #systemGetProperty}). <br>
+ * This behaviour can be disabled/changed by overriding these "system..." methods. <br>
+ * <br>
+ * <h3>Config-Values Translation</h3>
+ * All config values (originated from a config-source or from {@link #systemGetProperty})
+ *  are "translated" using {@link #translateConfigValue} (which by default does nothing). <br>
+ * <br>
+ */
+public class ConfigManager implements Closeable {
+
+    /** See {@link #addProvider}. All access must be "synchronized" ! */
+    private final List<ConfigProvider> configProviders = new LinkedList<>();
+
+    /** See {@link #addConfigSource}. All access must be "synchronized" ! */
+    private final List<ConfigProvider.ConfigSource> configSources = new LinkedList<>();
+
+    /** See {@link #registerChangeCallback}. All access must be "synchronized" ! */
+    private final Set<Runnable> changeCallbacks = new HashSet<>();
+
+    /** The current config-entries */
+    private final Map<String, ConfigEntry> previousConfigs = new ConcurrentHashMap<>();
+
+    private final Thread reloadSourcesThread;
+    private boolean isClosed = false;
+
+    /**
+     * Whenever we reload a config-source - this is incremented.
+     * At all times, we know - per config-entry - at what generation it was last encountered.
+     */
+    private long lastReloadGeneration = 0;
+
+    private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+
+    /** Construct a NON-RELOADING config-manager */
+    public ConfigManager() {
+        // Add all standard providers.
+        init();
+
+        reloadSourcesThread = null;
+    }
+
+    /**
+     * Construct a RELOADING config-manager.
+     * The periodic reload interval is also configurable.
+     * Note that
+     */
+    public ConfigManager(String reloadConfigKey, long reloadDefaultValue, TimeUnit reloadTimeUnit) {
+        // Add all standard providers.
+        init();
+
+        // Periodically reload configs.
+        DynamicConfigDuration interval = new DynamicConfigDuration(this, reloadConfigKey, reloadDefaultValue, reloadTimeUnit);
+        reloadSourcesThread = new Thread(() -> {
+            while (!isClosed) {
+                try {
+                    interval.sleep(sleepMilliseconds -> Math.max(sleepMilliseconds, 1000L));   // never sleep less than 1 second - a safety mechanism to avoid 100% CPU loop
+                } catch (InterruptedException ignored) {
+                }
+                reloadAllConfigs();
+            }
+        });
+        reloadSourcesThread.setName("ConfigManagerReloader");
+        reloadSourcesThread.setDaemon(true);
+        reloadSourcesThread.start();
+    }
+
+    /** Overridable: This is called upon construction and add all standard providers */
+    protected void init() {
```

Comment:
Don't think there's a need to override init and there's no need for ConfigManager to know the providers loaded.
You can get a list of providers from a property and add them all as ConfigProvider

Reply:
Right, but I believe adding all standard providers makes using this package easier.
Note that these are only the providers - not actual stores (the stores themselves should be taken from a property)